bigip_plugin/utils.py

```diff
@@ -0,0 +1,77 @@
+########
+# Copyright (c) 2018 GigaSpaces Technologies Ltd. All rights reserved
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#        http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+#    * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#    * See the License for the specific language governing permissions and
+#    * limitations under the License.
+
+# Standard Imports
+import sys
+
+# Third Party Imports
+from cloudify import ctx
+from cloudify.exceptions import NonRecoverableError
+from cloudify.utils import exception_to_error_cause
+
+# local Imports
+from bigip_sdk import gtm
+
+
+RELATIONSHIP_TYPE_MANAGED_BY_GTM = (
+    'cloudify.bigip.gtm.relationships.managed_by_gtm'
+)
+
+GTM_PASSWORD = 'password'
+GTM_USER = 'user'
+GTM_HOST = 'host'
+
+GTM_KEYS = (
+    GTM_PASSWORD,
+    GTM_USER,
+    GTM_HOST
+)
+
+
+def generate_traceback_exception():
+    _, exc_value, exc_traceback = sys.exc_info()
+    response = exception_to_error_cause(exc_value, exc_traceback)
+    return response
+
+
+def _retrieve_gtm_connection(resource_instance):
+    for relationship in resource_instance.relationships:
+        if relationship.type == RELATIONSHIP_TYPE_MANAGED_BY_GTM:
+            return relationship.target
+
+
+def _retrieve_properties(resource_instance):
+    target = _retrieve_gtm_connection(resource_instance)
+    return target.node.properties
+
+
+def with_gtm_client(function):
+    def wrapper(**kwargs):
+        # This will fetch the properties of the GTM instance from the
+        # relationship of the node
+        try:
+            properties = _retrieve_properties(ctx.instance)
+            gmt_prop = {k: properties[k] for k in GTM_KEYS}
+            kwargs['igtm'] = gtm.GTM(**gmt_prop)
+            function(**kwargs)
+        except Exception as error:
```

Comment:
Do we want all exceptions to be treated equally?

Reply:
I thought about this, and since we have a single node that do all the operations, adding another exceptions handling to cause a retry or recoverable could lead to create another resources already created  which raise another exception which will cause another complexity of the system to handle already created resources.

Using multiple nodes save us from this issue, but the asked to combine all the operations in one node. and based on that I added the built it